homeassistant/components/notify/facebook.py

```diff
@@ -115,3 +123,102 @@ def log_error(response):
     _LOGGER.error(
         "Error %s : %s (Code %s)", response.status_code, error_message,
         error_code)
+
+
+class FacebookReceiver(HomeAssistantView):
+    """Handle webhooks from Facebook."""
+
+    requires_auth = False
+    url = FACEBOOK_HANDLER_URL
+    name = 'api:facebook_webhooks'
+
+    def __init__(self, hass, config):
+        """Initialize the class."""
+        self.hass = hass
+        self.token = config[CONF_PAGE_ACCESS_TOKEN]
+        self.allowed_chat_ids = config.get(CONF_ALLOWED_CHAT_IDS)
+
+    @asyncio.coroutine
+    def get(self, request):
+        """Accept the GET verification from facebook."""
+        data = request.query
+
+        if data.get("hub.mode") == "subscribe" and data.get("hub.challenge"):
+            if not data.get("hub.verify_token") == self.token:
+                return "token mismatch", 403
+
+        return data.get("hub.challenge"), 200
+
+    @asyncio.coroutine
+    def post(self, request):
+        """Accept the POST from facebook."""
+        try:
+            data = yield from request.json()
+        except ValueError:
+            return self.json_message('Invalid JSON specified',
+                                     HTTP_BAD_REQUEST)
+
+        for entry in data["entry"]:
+            for messaging_event in entry["messaging"]:
+
+                if messaging_event.get("message"):
+
+                    sender_id = messaging_event["sender"]["id"]
+                    message_text = messaging_event["message"]["text"]
+
+                    if message_text == "get my id":
+                        self.reply_with_id(sender_id)
+                        return "ok", 200
+
+                    if self.allowed_chat_ids is not None:
+                        if int(sender_id) in self.allowed_chat_ids:
+                            self.process_message_event(sender_id,
+                                                       message_text,
+                                                       messaging_event)
+                        else:
+                            _LOGGER.warning(
+                                ("Received message on facebook webhook from "
+                                 "sender not in allowed chat ids."))
+                    else:
+                        _LOGGER.warning(
+                            "Recieved message but no allowed senders defined")
+
+        return "ok", 200
+
+    def process_message_event(self, sender_id, message_text, messaging_event):
+        """Process the incoming message event."""
+        _LOGGER.debug("Received facebook message.")
+        event_data = {
+            'sender_id': sender_id,
+            'message': message_text,
+            'payload': None
+        }
+
+        if "quick_reply" in messaging_event["message"]:
+            event_data['payload'] = \
+                messaging_event["message"]["quick_reply"]["payload"]
+
+        self.hass.bus.async_fire(EVENT_FACEBOOK_MESSAGE, event_data)
```

Comment:
Notify platforms should not handle webhooks. Turn this into a component.

Reply:
Ok, no problem! Should I move just the new functionality that handles the incoming messages to a new component and leave the existing notify platform alone?